### ISSUE TITLE
payload_offset() can panic on invalid data

### ISSUE THREAD

**Issue by ts252** [dholroyd/rtp-rs#1]
It needs to check that offset it reads the extension length from is within the buffer supplied. 

I do the following fix (which may not be to your taste):
```rust
    fn payload_offset(&self) -> usize {
        let offset = Self::MIN_HEADER_LEN + (4 * self.csrc_count()) as usize;
        if offset >= self.buf.len() - 3 {
            std::u16::MAX as usize
        } else if self.extension() {
            let len = (self.buf[offset + 2] as usize) << 8 | (self.buf[offset + 3] as usize);
            offset + 4 + len
        } else {
            offset
        }
    }
```

**Comment by dholroyd** [comment]
Thanks for pointing this out!

Looks like I am also not calculating the length correctly -- spec says it's a count of 32-bit units, and my current code assumed a count of 8-bit units.

Will try to rework this area - might as well try to add extension header support while we're at it.

(Did you notice this because if packets that actually have an extension, or just through malformed data?  If you actually had an example packet with legitimate payload, it would make a good test case!)

**Comment by dholroyd** [comment]
I did the fix in a different way, by adding extension header support -- if you are able to validate this change still fixes you problem that would be fantastic.

**Comment by ts252** [comment]
Yes this looks good to me! It handles all my test data.

I'm afraid I am using your code to speculatively parse all udp packets in a network capture, looking for RTP flows. I don't, therefore, have an example of a valid packet with an exotic header extension for you, just gigabytes of nonsense.

Anyway, thanks for creating and maintaining this useful crate.

**Comment by dholroyd** [comment]
Released in v0.4 - https://crates.io/crates/rtp-rs/0.4.0
